### ISSUE TITLE
Inconsistent handling of complex functions

### ISSUE THREAD

**Issue by j-towns** [HIPS/autograd#179]
My complex analysis is a bit rusty, and I'm getting confused by the handling of complex functions. Many functions are differentiable as complex functions (i.e. they are [holomorphic](https://en.wikipedia.org/wiki/Holomorphic_function)) and their complex derivatives are implemented in autograd.

However there also seem to be functions, like `abs`, `var`, `angle` and others, which are not differentiable as complex functions, but they also have derivatives implemented. I'm assuming these derivatives treat the complex inputs as if they were 2D real valued vectors? This seems inconsistent to me.

Users can fairly easily replicate the second behaviour without these derivatives being defined, by manually decomposing their numbers into real and imaginary parts, so I would tentatively propose removing these pseudo-derivatives...

Apologies if I'm making some dumb mistake here...

**Comment by j-towns** [comment]
I'm also surprised the tests pass in both of cases, which is making me think I must be missing something.

**Comment by dougalm** [comment]
Hi James, thanks for bringing this up. It's something we thought through carefully but never properly documented. I'll have a go here. Consider a complex-to-complex function, `f`, expressed in terms of real-to-real components, `u` and `v`:

	def f(z):
	    x, y = real(z), imag(z)
	    return u(x, y) + v(x, y) * 1j

We *define* `grad(f)` as

	def grad_f(z):
	    x, y = real(z), imag(z)
	    return grad(u, 0)(x, y) - i * grad(u, 1)(x, y)

(The second argument of `grad` specifies which argument we're differentiating with respect to.)

So we throw out `v`, the imaginary part of f, entirely. And that's it.

Why does this make sense? Well, it covers the two important cases:
  * If `f` is holomorphic, then this gives the usual complex derivative of a holomorphic function (since `grad(u, 0) == grad(v, 1)` and `grad(u, 1) == - grad(v, 0)`).
  * If `f` is a real-valued loss function of a complex parameter, `x`, then it gives a result that you can use in a gradient-based optimizer, by taking steps in the direction of the conjugate of `grad(f)(x)` (not to be confused with the conjugate gradient method!). 

Since optimization is Autograd's main intended application, I feel strongly that we should support this second case, even though such an `f` would be non-holomorphic. Maybe you want to compute a power spectrum as part of your objective function, which involves taking the `abs` of some FFT. As long as your output is scalar, you should be able to use Autograd to ask "how do the inputs of this function affect its output".

Where our convention fails is if you have some non-holomorphic function and you care about all of du/dx, du/dy, dv/dx and dv/dy. But then the answer would have to contain four real values and there would be no way to express it as a single complex number. A guiding principle throughout the development of autograd has always been that evaluating the gradient should give a result of the same type and shape as the input.

Chapter 4 of [my PhD thesis](https://dougalmaclaurin.com/phd-thesis.pdf) goes into a bit more detail about how we define the primitive vector-Jacobian products.

Does that make any sense?

**Comment by j-towns** [comment]
Awesome, yeah that makes perfect sense, thanks for the reply.

**Comment by j-towns** [comment]
I've done some scribbling and I think that the above approach won't work for forward mode, because the incoming forward gradient won't contain enough information to work out the next forward gradient...

However there is an analagous approach that will work, namely defining the forward derivative to be
```
def grad_f(z):
    x, y = real(z), imag(z)
    return grad(u, 0)(x, y) + i * grad(v, 0)(x, y)
```
I believe this will maintain the property that the derivative has the same type and shape as the _output_ (which makes sense for forward mode).

**Comment by rainwoodman** [comment]
Another possible convention is to say the real and imag part of a complex number are always treated as separate degrees of freedoms. The advantage of this convention is

- if the chain terminates at a complex value, then the gradient is not surprising. 

- gradient decent is along the direction of gradient (instead of the direction of the conjugate of the gradient)

- it probably makes sense for forward and backward model at the same time.

I suspect it will break the holomorphic properties (my complex analysis is pretty poor), but is there a case preserving the holomorphic properties is explicitly useful? 

I am not strongly advocating changing the convention in autograd, because it surely will break a lot of code.

**Comment by j-towns** [comment]
Reopening until we've documented this

**Comment by rainwoodman** [comment]
OK. Here is one more reason to treat complex numbers as tuples.

**Comment by j-towns** [comment]
I think I'm in favour of the current way things are implemented, partly because it would be a lot of work to change to the approach that you're suggesting. A lot of derivative implementations would become more complicated too, because using something that's compatible with the [standard complex derivative](http://mathworld.wolfram.com/ComplexDifferentiable.html) for holomorphic functions means that most derivatives extend to complex numbers without any need for a special case.

For example, at the moment the derivative of `y**2` is `2 * y`. If we used the approach that you're suggesting then the derivative of `y**2` would be `2 * np.conj(y)` for complex `y`, so you'd need to type check `y`. This is true for all standard nice math functions (which are holomorphic) — the derivative of any function composed of polynomials, exponentials, trig functions would need to be re-implemented to do type checking and to conjugate if the argument is complex.

**Comment by dougalm** [comment]
@j-towns - you're exactly right, and that's really the main reason we chose that convention in the first place. We realized that treating complex numbers as tuples would require type checking and `np.conj` in many of the primitive gradients, which would hurt performance and make the code ugly and complicated. I think it remains the right call.

**Comment by dougalm** [comment]
You're right that it should be documented. I'll add my explanation from earlier in this thread to the autograd tutorial.

**Comment by rainwoodman** [comment]
@j-towns:

The plain tuple is the same as (du/dx, dv/dx), or using the first column instead first row of the Jacobian. So I suspect if everything is transposed it would work out just fine.

@douglam, it appears that TenserFlow uses the du/dx, dv/dx convention. [1][2] Perhaps it is worthy mentioning the difference in the tutorial? 

[1] : Their vjp operator of FFT appears to be iFFT (identical to my derivations): https://github.com/tensorflow/tensorflow/blob/94f2229c9e4b4324a324330c8f419276eda7e503/tensorflow/python/ops/math_grad.py#L961 
[2] : Their vjp operator of Pow appears to be doing exactly the conjugate: https://github.com/tensorflow/tensorflow/blob/94f2229c9e4b4324a324330c8f419276eda7e503/tensorflow/python/ops/math_grad.py#L678

**Comment by j-towns** [comment]
By the way, it's trivial to write a wrapper that computes conjugate derivatives, and maybe this could be put in `convenience_wrappers.py`.

```python
In [1]: from autograd import grad
   ...: import autograd.numpy as np
   ...: 
   ...: def grad_conj(fun, argnum=0):
   ...:     return lambda *args, **kwargs: np.conj(grad(fun, argnum)(*args, **kwargs))
   ...: 

In [2]: grad(np.exp)(2 + 1j)
Out[2]: (3.9923240484412719+6.2176763123679679j)

In [3]: grad_conj(np.exp)(2 + 1j)
Out[3]: (3.9923240484412719-6.2176763123679679j)
```

**Comment by rainwoodman** [comment]
Sure. The docstring of `grad_conj` could be the place to add the discussion on complex number conventions. It is unfortunate I could not find any direct evidence of TensorFlow's treatment in their documentation / paper.
